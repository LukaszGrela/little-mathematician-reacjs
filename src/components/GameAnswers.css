.game-answers {
    width: 300px;
    margin: 0 auto;
}
.answer {
    padding: 20px;
    margin: 2px;

    min-width: 70px;
    border: 1px solid #ccc;
}

.answer:focus {
    outline: none;
}