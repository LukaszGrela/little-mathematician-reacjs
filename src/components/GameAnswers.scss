/*
   Copyright 2018 ≈Åukasz 'Severiaan' Grela GrelaDesign

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

@import '../sass/_variables.scss';
.game-answers {
    width: $width-game-column;
    margin: 0 auto;
    
    button {
        /* reset */
        background-color: $primary-color-text;
    }
    hr {
        border-color: $divider-color;
        border-style: solid;
    }
}

.answer {
    padding: 20px;
    margin: 2px;
    min-width: 70px;
    border: 1px solid $divider-color;
    font-size: 1em;
    &:focus {
        outline: none;
    }
}


.hidden {
    visibility: hidden;
}