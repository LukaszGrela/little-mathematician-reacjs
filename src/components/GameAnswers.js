import React, { Component } from 'react';

import './GameAnswers.css'

class GameAnswers extends Component {
    answered = false;
    constructor(props) {
        super(props);
        this.state = {};
        console.log("GameAnswers", props);
    }



    shouldComponentUpdate(nextProps, nextState) {

        // do not render when options hasn't changed
        if (nextProps.options && this.compare(nextProps.options, this.props.options)) return false;

        return true;
    }

    compare(arrA, arrB) {
        if (!arrA || !arrB) return false;

        if (arrA.length !== arrB.length) return false;

        for (let index = 0; index < arrA.length; index++) {
            const elementA = arrA[index];
            if (elementA !== arrB[index]) return false;
        }

        return true;
    }

    shuffle(array) {
        let m = array.length, t, i;

        // While there remain elements to shuffle…
        while (m) {

            // Pick a remaining element…
            i = Math.floor(Math.random() * m--);

            // And swap it with the current element.
            t = array[m];
            array[m] = array[i];
            array[i] = t;
        }

        return array;
    }



    generateAnswers() {
        let i = 0;
        let options = this.props.options;

        options = this.shuffle(options);

        this.answered = false;
        return options.map((option) => <button className="answer"
            key={i++}
            onClick={(e) => {
                if (this.answered) return;
                const button = e.target;
                if (button.className.indexOf('hidden') === -1)
                    button.className += ' hidden';

                this.props.onSelection && this.props.onSelection(option, i);
                this.answered = true;
            }}>{option}</button>);
    }

    render() {
        return (
            <div className="game-answers">
                <hr />
                {
                    this.generateAnswers()
                }
            </div>
        );
    }
}

export default GameAnswers;