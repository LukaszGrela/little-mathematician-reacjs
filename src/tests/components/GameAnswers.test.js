import React from 'react';
import { shallow } from 'enzyme';
import GameAnswers from '../../components/GameAnswers';


let wrapper, onSelection, options, selectionId;
beforeEach(() => {
    onSelection = jest.fn();
    options = [4, 3, 2, 1];
    wrapper = shallow(<GameAnswers
        selectionId={selectionId}
        onSelection={onSelection}
        options={options} />);
});
test('Should GameAnswers to render correctly', () => {
    expect(wrapper).toMatchSnapshot();
});
// 
test('Should render corrent number of options', () => { 
    const buttons = wrapper.find('.answer');
    expect(buttons).toHaveLength(options.length);
});
// 
test('Should hide selected option', () => { 
    options.forEach((i,index) => {
        let selectionId = index + 1;
        wrapper.setProps({selectionId});
        let hidden = wrapper.find('.answer.hidden');
        expect(hidden).toHaveLength(1);
    });
});
// 
test('Should call onSelection with correct arguments', () => { 
    options.forEach((i,index) => {
        let id = index+1;
        let className = 'option-' + id;
        let button = wrapper.find('.answer.'+className);
        expect(button).toHaveLength(1);
        button.simulate('click');
        expect(onSelection).toHaveBeenLastCalledWith(i,id);
    }); 
});
//
test('Should GameAnswers className have locked when selectionId exists', () => {

    let locked = wrapper.find('.game-answers.locked');
    expect(locked).toHaveLength(0);    

    let selectionId = 1;
    wrapper.setProps({selectionId});
    locked = wrapper.find('.game-answers.locked');
    expect(locked).toHaveLength(1);    
});