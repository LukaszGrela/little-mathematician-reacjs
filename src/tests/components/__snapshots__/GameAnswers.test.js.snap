// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should GameAnswers to render correctly 1`] = `
<div
  className="game-answers"
>
  <hr />
  <button
    className="answer option-1"
    key="1"
    onClick={[Function]}
  >
    4
  </button>
  <button
    className="answer option-2"
    key="2"
    onClick={[Function]}
  >
    3
  </button>
  <button
    className="answer option-3"
    key="3"
    onClick={[Function]}
  >
    2
  </button>
  <button
    className="answer option-4"
    key="4"
    onClick={[Function]}
  >
    1
  </button>
</div>
`;
